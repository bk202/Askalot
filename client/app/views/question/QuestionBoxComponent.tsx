import * as React from "react";
import {Component} from "react";
import {connect} from "react-redux";
import {AppStoreState} from "../../stores/AppStore";
import {QuestionActions} from "../../actions/QuestionActions";
import {QuestionHomeReducerState} from "../../reducers/QuestionHomeReducer";
import {ErrorReducerState} from "../../reducers/ErrorReducer";
import {FrontEndQuestionModels} from "../../models/QuestionModels";
import QuestionPreview = FrontEndQuestionModels.QuestionPreview;

export interface QuestionViewProps extends QuestionHomeReducerState {
    loggedIn: boolean;
    globalError: ErrorReducerState;
    fetchQuestion: () => void;
}

class QuestionView extends Component<QuestionViewProps, any> {

    constructor(props) {
        super(props);
    }

    render() {
        return (
            <div>
            </div>
        )
    }
}