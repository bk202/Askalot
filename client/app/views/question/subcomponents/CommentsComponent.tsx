import * as React from "react";
import {CommentDto} from "../../../../../server/dtos/q&a/CommentDto";
import PropTypes from 'prop-types';
import {withStyles, createStyleSheet} from 'material-ui/styles';
import List, {ListItem, ListItemText} from 'material-ui/List';
import Divider from 'material-ui/Divider';
import Icon from 'material-ui/Icon';
import {CSSProperties} from "react";
import TextField from 'material-ui/TextField';
import Button from 'material-ui/Button';
import IconButton from 'material-ui/IconButton'
import {UserDto} from "../../../../../server/dtos/auth/UserDto";

export interface CommentsComponentProps {
    comments: CommentDto[];
    user : UserDto;
    onCommentsSubmit : (comments: CommentDto[]) => void;
}

export interface CommentsComponentState {
    inputMode: boolean;
    commentContent: string;
    errorMsg: string;
}

const styleSheet: CSSProperties = {
    root: {
        width: '100%',
        maxWidth: '360px'
    }
};

export class CommentsComponent extends React.Component<CommentsComponentProps, CommentsComponentState> {
    constructor(props) {
        super (props);
        this.state = {
            inputMode: false,
            commentContent: "",
            errorMsg: "",
        };
    }


    addNewComment = () => {
        if(this.state.commentContent){
            this.setState({errorMsg: ""});

        }
        else{
            this.setState({errorMsg: "Cannot submit empty comment."});
        }
    }

    onCommentChange = (event)  => {
        this.setState({commentContent: event.target.value});
    }

    onErrMsg = () => {
        if(this.state.errorMsg != ""){
            return(
                <p><mark>{this.state.errorMsg}</mark></p>
            )
        }
    }

    renderInputCommentBox() {
        if (this.state.inputMode) {
            return (
                <div>
                    {this.onErrMsg()}
                    <form>
                        <textarea rows={4} cols={50} value={this.state.commentContent} onChange={this.onCommentChange}></textarea>
                    </form>
                    <Button raised color="primary" className={styleSheet.root} onClick={this.addNewComment}>
                        Submit
                    </Button>
                </div>
            )
        }
    }

    render() {
        return (
            <div>
                <List className={"Comments"}>
                    <ListItem button>
                        <ListItemText primary="Inbox"/>
                    </ListItem>
                    <Divider light/>
                    <ListItem button>
                        <ListItemText primary="Drafts"/>
                    </ListItem>
                    <Divider />
                    <ListItem button>
                        <ListItemText primary="Trash"/>
                    </ListItem>
                </List>
                <div className={styleSheet.root}>
                    <IconButton>
                    <Icon onClick={() => this.setState({inputMode: !this.state.inputMode}) }>add_circle</Icon>
                    </IconButton>
                </div>
                {this.renderInputCommentBox()}
            </div>
        );
    }
}