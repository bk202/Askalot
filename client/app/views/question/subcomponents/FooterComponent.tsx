import * as React from "react";
import {UserDto} from "../../../../../server/dtos/auth/UserDto";
import {ThumbComponent} from "../../../components/Forms/ThumbComponent";

interface props {
    onUpVote: () => void;
    onDownVote: () => void;
    upVotes: number;
    downVotes: number;
    author: UserDto;
    createdUtc: Date;
}

export class FooterComponent extends React.Component<props> {


    render() {
        let question = {
            upVotes: 0
        };
        return (
            <div>
                <ThumbComponent
                    value={this.props.upVotes}
                    thumbUp={true}
                    onClick={this.props.onUpVote} />
                <ThumbComponent
                    value={this.props.downVotes}
                    thumbUp={false}
                    onClick={this.props.onDownVote} />
                <div style={{color: "grey", fontSize: 10, textAlign: "right"}}>
                    {this.props.createdUtc && <div>Posted on {this.props.createdUtc}</div>}
                    <br/>
                    by {this.props.author.username}
                </div>
            </div>
        );
    }
}