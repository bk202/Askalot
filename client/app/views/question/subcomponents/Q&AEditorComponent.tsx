import * as React from "react";
import {Component} from "react";
import {FrontEndQuestionModels} from "../../../models/QuestionModels";
import {EditorState} from "draft-js";
import {CustomEditor} from "../../../components/CustomEditor/CustomEditor";
import Button from "material-ui/Button";
import QuestionPreview = FrontEndQuestionModels.QuestionPreview;

export interface QAEditorProps{
    value: EditorState;
    onChange: (state: EditorState) => any;
    onSubmit: () => any;
    readOnly?: boolean;
    style?:any;
}

export class QAEditorComponent extends Component<QAEditorProps> {
    render() {
        const readOnly = this.props.readOnly;
        return (
            <div style={{marginBottom: 15}}>
                <CustomEditor value={this.props.value} onChange={this.props.onChange} readOnly={readOnly}
                            border={!readOnly} style={this.props.style}
                />
                {
                    !this.props.readOnly &&
                    <div style={{textAlign: "right"}}>
                        <Button onClick={this.props.onSubmit}>save</Button>
                    </div>
                }
            </div>
        )
    }
}