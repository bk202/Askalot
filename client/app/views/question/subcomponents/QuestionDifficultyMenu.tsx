import * as React from "react";
import {DropDownSelect} from "../../../components/DropDownSelect";
import {DifficultyLevel, QuestionEducationLevel} from "../../../../../server/enums/QuestionEducationLevel";
import {getDropDownDataFromNumericalEnum, getDropDownDataFromStringEnum} from "../../../utils/utils";
import {QuestionDifficulty} from "../../../../../server/models/Question";

interface props {
    onDifficultyChange: (diff: QuestionDifficulty) => void;
    difficulty: QuestionDifficulty;
}

export class QuestionDifficultyMenu extends React.Component<props> {
    render() {
        return (
            <div>
                <DropDownSelect
                    placeholder="Question Level"
                    data={getDropDownDataFromStringEnum(QuestionEducationLevel)}
                    onChange={(educationLevel) => this.props.onDifficultyChange({
                        educationLevel, difficultyLevel: this.props.difficulty.difficultyLevel
                    })}
                    defaultValue={QuestionEducationLevel.NOT_SPECIFIED}
                />
                { this.props.difficulty.educationLevel != QuestionEducationLevel.NOT_SPECIFIED &&
                <DropDownSelect
                    placeholder="Difficulty Level"
                    data={getDropDownDataFromNumericalEnum(DifficultyLevel)}
                    onChange={(difficultyLevel) => this.props.onDifficultyChange({
                        educationLevel: this.props.difficulty.educationLevel, difficultyLevel
                    })}
                    defaultValue={DifficultyLevel.NOT_SPECIFIED}
                />}
            </div>
        );
    }
}