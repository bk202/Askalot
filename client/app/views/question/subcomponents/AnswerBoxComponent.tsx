import * as React from "react";
import {Component} from "react";
import {FrontEndQuestionModels} from "../../../models/QuestionModels";
import Divider from "material-ui/Divider";
import {QAEditorComponent} from "./Q&AEditorComponent";
import {UserDto} from "../../../../../server/dtos/auth/UserDto";
import Button from "material-ui/Button";
import Paper from "material-ui/Paper";
import Grid from "material-ui/Grid";
import QuestionPreview = FrontEndQuestionModels.QuestionPreview;
import Question = FrontEndQuestionModels.Question;
import Answer = FrontEndQuestionModels.Answer;

export interface AnswerBoxComponentProps {
    onAnswerChange: (answer: Answer) => void;
    onSubmit: () => void;
    onEditClick: () => void;
    user: UserDto;
    answer: Answer;
    editMode: boolean;
    resetAnswer?: () => void;
}

const paperStyle = {height: "100%", padding: 15, paddingBottom: 0};

export class AnswerBoxComponent extends Component<AnswerBoxComponentProps> {

    onContentChange = (editorState) => {
        let question = this.props.answer;
        question.content = editorState;
        this.props.onAnswerChange(question);
    };

    render() {
        const answer = {...this.props.answer};
        const editable = (this.props.user && this.props.user.username === answer.author.username);
        return (
            <Paper style={paperStyle}>
                <div>
                    <Grid container justify="flex-end">
                        <Grid item>
                            {editable && !this.props.editMode && <Button onClick={this.props.onEditClick}>Edit</Button>}
                        </Grid>
                    </Grid>
                    <QAEditorComponent value={this.props.answer.content} onChange={this.onContentChange}
                                       onSubmit={this.props.onSubmit} readOnly={!this.props.editMode}
                                       style={{fontSize: 14}} reset={this.props.resetAnswer}
                    />
                    <Divider/>
                    <p style={{color: "grey", fontSize: 10, textAlign: "right"}}>
                        {answer.createdUtc && <div>Posted on {answer.createdUtc}</div>}
                        <br/>by {answer.author.username}
                    </p>
                </div>
            </Paper>
        )
    }
}