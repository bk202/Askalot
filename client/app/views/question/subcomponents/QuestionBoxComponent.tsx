import * as React from "react";
import {Component} from "react";
import {FrontEndQuestionModels} from "../../../models/QuestionModels";
import {EditableMultiPurposeHeader} from "../../../components/Headers/EditableMultiPurposeHeader";
import {ChipListComponent} from "../../../components/ChipListComponent";
import Divider from "material-ui/Divider";
import {QAEditorComponent} from "./Q&AEditorComponent";
import {UserDto} from "../../../../../server/dtos/auth/UserDto";
import Button from "material-ui/Button";
import Paper from "material-ui/Paper";
import QuestionPreview = FrontEndQuestionModels.QuestionPreview;
import Question = FrontEndQuestionModels.Question;
import cloneQuestion = FrontEndQuestionModels.cloneQuestion;

export interface QuestionBoxComponentProps {
    onQuestionChange: (question: Question) => void;
    onSubmit: () => void;
    onEditClick?: () => void;
    user: UserDto;
    question: Question;
    editMode: boolean;
    resetQuestion? : () => void ;
}

let paperStyle = {height: "100%", padding: 15, paddingBottom: 0};

export class QuestionBoxComponent extends Component<QuestionBoxComponentProps> {
    onTitleChange = (event) => {
        let question = cloneQuestion(this.props.question);
        question.title = event.target.value;
        this.props.onQuestionChange(question);
    };

    onContentChange = (editorState) => {
        let question = cloneQuestion(this.props.question);
        question.content = editorState;
        this.props.onQuestionChange(question);
    };

    render() {
        const question = {...this.props.question};
        const editable = this.props.user.username === question.author.username;
        let editButton;
        if (editable && !this.props.editMode){
            editButton = (
                <div style={{float: "right"}}>
                    <Button color="primary" onClick={this.props.onEditClick}>Edit</Button>
                </div>
            )
        }
        return (
            <Paper style={paperStyle}>
                <EditableMultiPurposeHeader value={question.title} editMode={this.props.editMode}
                                            onEditClick={this.props.onEditClick}
                                            onTitleChange={this.onTitleChange}/>
                {editButton}
                <Divider />
                <div>
                    <QAEditorComponent value={this.props.question.content} onChange={this.onContentChange}
                                       onSubmit={this.props.onSubmit} readOnly={!this.props.editMode}
                                        style={{fontSize: 14}} reset={this.props.resetQuestion}
                    />
                    <div>
                        <ChipListComponent chips={question.tags} keyName={"tag"}/>
                    </div>
                    <Divider/>
                    <p style={{color: "grey", fontSize: 10, textAlign: "right"}}>
                        Posted on {question.createdUtc}<br/>by {question.author.username}
                    </p>
                </div>
            </Paper>
        )
    }
}