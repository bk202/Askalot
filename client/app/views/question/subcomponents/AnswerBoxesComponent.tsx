import * as React from "react";
import {Component} from "react";
import {FrontEndQuestionModels} from "../../../models/QuestionModels";
import {UserDto} from "../../../../../server/dtos/auth/UserDto";
import {AnswerBoxComponent, AnswerBoxView} from "./AnswerBoxComponent";
import Button from "material-ui/Button";
import AddIcon from "material-ui-icons/Add";
import QuestionPreview = FrontEndQuestionModels.QuestionPreview;
import Question = FrontEndQuestionModels.Question;
import Answer = FrontEndQuestionModels.Answer;
import cloneAnswers = FrontEndQuestionModels.cloneAnswers;
import {connect} from "react-redux";
import {AppStoreState} from "../../../stores/AppStore";
import {AnswerActions} from "../../../actions/AnswerActions";
import {QuestionPageReducerState} from "../../../reducers/QuestionPageReducer";
import cloneAnswer = FrontEndQuestionModels.cloneAnswer;
import {Prompt} from "react-router";
export interface AnswerBoxesComponentProps {
}

interface props extends AnswerBoxesComponentProps, DispatchToProps, StateToProps {
}

export interface AnswerBoxesComponentState {
    answerId: string;
    editAnswer: boolean;
    answers: Answer[];
}

export class AnswerBoxesComponent extends Component<props, AnswerBoxesComponentState> {
    state = {answerId: undefined, editAnswer: false, answers: []};

    componentWillMount() {
        this.setState({answers: cloneAnswers(this.props.answers)});
    }

    componentWillReceiveProps(nextProps: props) {
        if (nextProps.answers != this.props.answers) {
            this.setState({editAnswer: false, answers: cloneAnswers(nextProps.answers)});
        }
    }

    componentWillUnmount() {
        this.resetAnswers();
    }

    onEditClick = (answer: Answer) => {
        this.setState({answerId: answer._id, editAnswer: true})
    };

    onAnswerChange = (answer: Answer) => {
        let answers = this.state.answers;
        answers = answers.map((ans) => {
            ans = cloneAnswer(ans);
            if (answer._id === ans._id) {
                return answer;
            }
            return ans;
        });
        this.setState({answers});
    };

    addAnswer = () => {
        let answers = cloneAnswers(this.state.answers);
        let answer = new Answer(this.props.question._id, this.props.user);
        answers.push(answer);
        this.setState({answerId: answer._id, editAnswer: true, answers});
    };

    resetAnswers = () => {
        this.setState({editAnswer: false, answerId: undefined});
        let answers = cloneAnswers(this.props.answers);
        this.setState({answers});
    };

    onSubmit = (answer: Answer) => {
        if (answer._id) {
            this.props.editAnswer(answer);
        } else {
            this.props.addAnswer(answer);
        }
    };

    render() {
        return (
            <div>
                <Prompt
                    when={this.state.editAnswer}
                    message={location => (
                        `All unsaved changes will be discarded. Are you sure you want to leave?`
                    )}
                />
                {this.state.answers.map((answer) => {
                    let key = (answer._id) ? answer._id : "new question key";
                    const editMode = this.state.editAnswer && (answer._id === this.state.answerId || answer._id === undefined);
                    return (
                        <AnswerBoxView editMode={editMode}
                                       user={this.props.user}
                                       onSubmit={() => this.onSubmit(answer)}
                                       answer={answer}
                                       onAnswerChange={this.onAnswerChange}
                                       key={key}
                                       onEditClick={() => this.onEditClick(answer)}
                                       resetAnswer={this.resetAnswers}
                        />
                    )
                })}
                {(this.props.user && !this.state.editAnswer) &&
                <div key="edit-answer-button" style={{float: 'right', marginTop: 5}}>
                    <Button fab color="primary" onClick={this.addAnswer}>
                        <AddIcon/>
                    </Button>
                </div>}

            </div>
        )
    }
}

interface  StateToProps {
    user: UserDto
    question: Question;
    answers: Answer[];
}

interface DispatchToProps {
    editAnswer: (answer: Answer) => void;
    addAnswer: (answer: Answer) => void;
    upVoteAnswer: (answer: Answer) => void;
    downVoteAnswer: (answer: Answer) => void;
}

const mapStateToProps = (state: AppStoreState): StateToProps => ({
    user: state.auth.user,
    question: state.questionPage.questionPage.question,
    answers: state.questionPage.questionPage.answers
});

const mapDispatchToProps = (dispatch): DispatchToProps => ({
    editAnswer: (answer: Answer) => dispatch(AnswerActions.updateAnswer(answer)),
    addAnswer: (answer: Answer) => dispatch(AnswerActions.createAnswer(answer)),
    upVoteAnswer: (answer: Answer) => dispatch(AnswerActions.createAnswer(answer)),
    downVoteAnswer: (answer: Answer) => dispatch(AnswerActions.createAnswer(answer))
});

export const AnswerBoxesView = connect<StateToProps, DispatchToProps, any>(
    mapStateToProps, mapDispatchToProps)(AnswerBoxesComponent);