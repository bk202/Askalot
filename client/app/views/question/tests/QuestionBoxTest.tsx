import {QuestionBoxComponent} from "../subcomponents/QuestionBoxComponent";
import * as React from "react";
import * as renderer from 'react-test-renderer';
import {UserDto} from "../../../../../server/dtos/auth/UserDto";
import {UserTypeEnum} from "../../../../../server/enums/UserTypeEnum";
import {FrontEndQuestionModels} from "../../../models/QuestionModels";
import Question = FrontEndQuestionModels.Question;
import {Route, Router, withRouter} from "react-router";
import {RouterController} from "../../../api.controllers/RouterController";
import {MuiThemeProvider, createMuiTheme} from "material-ui/styles";
import {CommentDto} from "../../../../../server/dtos/q&a/CommentDto";


jest.mock('draft-js/lib/generateRandomKey', () => () => '123');
jest.mock('../../../components/CustomEditor/EditorFactory');
const e = require("../../../components/CustomEditor/EditorFactory");
const mMock = jest.fn();
e.EditorFactory.mockImplementation(() => {
    return {
        createRichEditor: () => null
    }
});

let muiTheme = createMuiTheme();
test('render question', () => {
    let user: UserDto = {
        _id: "something",
        company: undefined,
        email: "sheldon@email.com",
        points: 0,
        role: UserTypeEnum.NORMAL,
        university: undefined,
        username: "sheldon",
        verified: true,
        country: undefined
    };
    const functionStub = () => {
    };
    let questionStub: Question = new Question();
    questionStub.author = user;
    let propsStub = {
    //     user: UserDto; // current user
    // question: Question; // original question
    // questionEditorState: Question; // editor state
    // edit: boolean; // edit mode
        user: user,
        question: questionStub,
        questionEditorState: questionStub,
        edit: true,
        upVoteQuestion: functionStub,
        downVoteQuestion: functionStub,
        editQuestion: functionStub,
        changeQuestionEditorState: functionStub,
        createComment: (c: CommentDto, id: string) => {},
        updateComment: (c: CommentDto, id: string) => {},
        deleteComment: (c: CommentDto, id: string) => {}
        /*
            upVoteQuestion: (question: Question) => void;
            downVoteQuestion: (question: Question) => void;
            editQuestion: (question: Question) => void;
            changeQuestionEditorState: (state: QuestionEditorReducerState) => void;
            createComment: (c: CommentDto, id: string) => void;
            updateComment: (c: CommentDto, id: string) => void;
            deleteComment: (c: CommentDto, id: string) => void;*/
    };

    const component = renderer.create(
        <MuiThemeProvider theme={muiTheme}>
        <Router history={RouterController.history}>
            <Route render={(props: any) => {
                return <QuestionBoxComponent {...propsStub} {...props} />
            }} />
        </Router>
        </MuiThemeProvider>);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});