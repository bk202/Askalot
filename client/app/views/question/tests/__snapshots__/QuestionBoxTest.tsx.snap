// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render question box not logged in 1`] = `
<div>
  <div
    className="MuiPaper-root-1 MuiPaper-shadow0-3 MuiPaper-rounded-2"
    style={
      Object {
        "height": "100%",
        "padding": 15,
      }
    }
  >
    <div
      className="MuiFormControl-root-28"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-37 MuiInputLabel-root-32 MuiInputLabel-formControl-33 MuiInputLabel-animated-35"
        htmlFor={undefined}
      >
        Title
      </label>
      <div
        className="MuiInput-root-41 MuiInput-formControl-42 MuiInput-inkbar-43 MuiInput-underline-49"
      >
        <input
          aria-required={undefined}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-45 MuiInput-inputSingleline-52"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          rows={undefined}
          type="text"
          value=""
        />
      </div>
    </div>
    <hr
      className="MuiDivider-root-55 MuiDivider-default-56"
    />
    <div>
      <div
        style={
          Object {
            "marginBottom": 15,
          }
        }
      >
        <div
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          <button
            className="MuiButtonBase-root-74 MuiButton-root-60 MuiButton-flatPrimary-63"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-62"
            >
              save
            </span>
            <span
              className="MuiTouchRipple-root-76"
            />
          </button>
          <button
            className="MuiButtonBase-root-74 MuiButton-root-60"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-62"
            >
              cancel
            </span>
            <span
              className="MuiTouchRipple-root-76"
            />
          </button>
        </div>
      </div>
      <div>
        <div />
      </div>
      <hr
        className="MuiDivider-root-55 MuiDivider-default-56"
      />
      <div>
        <div
          className="MuiGrid-typeContainer-83 MuiGrid-spacing-xs-16-98"
        >
          <div
            className="MuiGrid-typeItem-84 MuiGrid-grid-xs-6-107"
          >
            <div
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <button
                className="MuiButtonBase-root-74 MuiIconButton-root-166"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                role="button"
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiIconButton-label-172"
                >
                  <span
                    aria-hidden="true"
                    className="material-icons MuiIcon-root-175 MuiIconButton-icon-173"
                  >
                    thumb_up
                  </span>
                  <span
                    style={
                      Object {
                        "fontSize": 14,
                        "marginLeft": 5,
                      }
                    }
                  />
                </span>
                <span
                  className="MuiTouchRipple-root-76"
                />
              </button>
            </div>
            <div
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <button
                className="MuiButtonBase-root-74 MuiIconButton-root-166"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                role="button"
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiIconButton-label-172"
                >
                  <span
                    aria-hidden="true"
                    className="material-icons MuiIcon-root-175 MuiIconButton-icon-173"
                  >
                    thumb_down
                  </span>
                  <span
                    style={
                      Object {
                        "fontSize": 14,
                        "marginLeft": 5,
                      }
                    }
                  />
                </span>
                <span
                  className="MuiTouchRipple-root-76"
                />
              </button>
            </div>
          </div>
          <div
            className="MuiGrid-typeItem-84 MuiGrid-grid-xs-6-107"
          >
            <div
              style={
                Object {
                  "paddingTop": 5,
                  "textAlign": "right",
                }
              }
            >
               â€“ 
              <p
                className="MuiTypography-root-182 MuiTypography-body1-191"
                style={
                  Object {
                    "color": "#1f3b99",
                    "display": "inline-block",
                  }
                }
              >
                sheldon
              </p>
              <span
                className="MuiTypography-root-182 MuiTypography-caption-192"
                style={
                  Object {
                    "display": "inline-block",
                    "marginLeft": 5,
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-typeContainer-83 MuiGrid-spacing-xs-16-98 MuiGrid-justify-xs-flex-end-94"
  >
    <div
      className="MuiGrid-typeItem-84 MuiGrid-grid-xs-12-113"
      style={
        Object {
          "background": "white",
          "paddingLeft": 15,
          "paddingRight": 15,
        }
      }
    >
      <ul
        className="MuiList-root-204 MuiList-padding-205 Comments"
      />
    </div>
  </div>
</div>
`;

exports[`render question logged in 1`] = `
<div>
  <div
    className="MuiPaper-root-1 MuiPaper-shadow0-3 MuiPaper-rounded-2"
    style={
      Object {
        "height": "100%",
        "padding": 15,
      }
    }
  >
    <div
      className="MuiFormControl-root-28"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <label
        className="MuiFormLabel-root-37 MuiInputLabel-root-32 MuiInputLabel-formControl-33 MuiInputLabel-animated-35"
        htmlFor={undefined}
      >
        Title
      </label>
      <div
        className="MuiInput-root-41 MuiInput-formControl-42 MuiInput-inkbar-43 MuiInput-underline-49"
      >
        <input
          aria-required={undefined}
          autoComplete={undefined}
          autoFocus={undefined}
          className="MuiInput-input-45 MuiInput-inputSingleline-52"
          defaultValue={undefined}
          disabled={false}
          id={undefined}
          name={undefined}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={undefined}
          onKeyUp={undefined}
          placeholder={undefined}
          rows={undefined}
          type="text"
          value=""
        />
      </div>
    </div>
    <hr
      className="MuiDivider-root-55 MuiDivider-default-56"
    />
    <div>
      <div
        style={
          Object {
            "marginBottom": 15,
          }
        }
      >
        <div
          style={
            Object {
              "textAlign": "right",
            }
          }
        >
          <button
            className="MuiButtonBase-root-74 MuiButton-root-60 MuiButton-flatPrimary-63"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-62"
            >
              save
            </span>
            <span
              className="MuiTouchRipple-root-76"
            />
          </button>
          <button
            className="MuiButtonBase-root-74 MuiButton-root-60"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex="0"
            type="button"
          >
            <span
              className="MuiButton-label-62"
            >
              cancel
            </span>
            <span
              className="MuiTouchRipple-root-76"
            />
          </button>
        </div>
      </div>
      <div>
        <div />
      </div>
      <hr
        className="MuiDivider-root-55 MuiDivider-default-56"
      />
      <div>
        <div
          className="MuiGrid-typeContainer-83 MuiGrid-spacing-xs-16-98"
        >
          <div
            className="MuiGrid-typeItem-84 MuiGrid-grid-xs-6-107"
          >
            <div
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <button
                className="MuiButtonBase-root-74 MuiIconButton-root-166"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                role="button"
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiIconButton-label-172"
                >
                  <span
                    aria-hidden="true"
                    className="material-icons MuiIcon-root-175 MuiIconButton-icon-173"
                  >
                    thumb_up
                  </span>
                  <span
                    style={
                      Object {
                        "fontSize": 14,
                        "marginLeft": 5,
                      }
                    }
                  />
                </span>
                <span
                  className="MuiTouchRipple-root-76"
                />
              </button>
            </div>
            <div
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <button
                className="MuiButtonBase-root-74 MuiIconButton-root-166"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
                role="button"
                tabIndex="0"
                type="button"
              >
                <span
                  className="MuiIconButton-label-172"
                >
                  <span
                    aria-hidden="true"
                    className="material-icons MuiIcon-root-175 MuiIconButton-icon-173"
                  >
                    thumb_down
                  </span>
                  <span
                    style={
                      Object {
                        "fontSize": 14,
                        "marginLeft": 5,
                      }
                    }
                  />
                </span>
                <span
                  className="MuiTouchRipple-root-76"
                />
              </button>
            </div>
          </div>
          <div
            className="MuiGrid-typeItem-84 MuiGrid-grid-xs-6-107"
          >
            <div
              style={
                Object {
                  "paddingTop": 5,
                  "textAlign": "right",
                }
              }
            >
               â€“ 
              <p
                className="MuiTypography-root-182 MuiTypography-body1-191"
                style={
                  Object {
                    "color": "#1f3b99",
                    "display": "inline-block",
                  }
                }
              >
                sheldon
              </p>
              <span
                className="MuiTypography-root-182 MuiTypography-caption-192"
                style={
                  Object {
                    "display": "inline-block",
                    "marginLeft": 5,
                  }
                }
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-typeContainer-83 MuiGrid-spacing-xs-16-98 MuiGrid-justify-xs-flex-end-94"
  >
    <div
      className="MuiGrid-typeItem-84 MuiGrid-grid-xs-12-113"
      style={
        Object {
          "background": "white",
          "paddingLeft": 15,
          "paddingRight": 15,
        }
      }
    >
      <ul
        className="MuiList-root-204 MuiList-padding-205 Comments"
      />
      <div
        style={
          Object {
            "height": 36,
            "marginBottom": 10,
          }
        }
      >
        <button
          className="MuiButtonBase-root-74 MuiButton-root-60 MuiButton-flatPrimary-63 MuiButton-dense-61"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchStart={[Function]}
          role="button"
          style={
            Object {
              "float": "right",
            }
          }
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-62"
          >
            Comment
          </span>
          <span
            className="MuiTouchRipple-root-76"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
